### ISSUE TITLE
[phpspec 3.1 beta] Method not implemented fatal errors not caught

### ISSUE THREAD

**Issue by mheki** [phpspec/phpspec#998]
I pretty much like new phpspec feature resulting with fatal errors marked as broken examples. I've found one issue - for some reason, method not implemented fatal errors are still red.

How to replicate?

```
<?php

interface Foo
{
    public function foo();
}
```

```
<?php

class Bar implements Foo
{
}
```

<img width="971" alt="screen shot 2016-09-07 at 08 00 23" src="https://cloud.githubusercontent.com/assets/2435655/18302735/445b4e66-74d1-11e6-8908-391d85b93cad.png">

**Comment by ciaranmcnulty** [comment]
Thanks @mheki - we currently catch parse errors when including the class and runtime errors when running the methods. I guess this is another case that happens somewhere inbetween!

**Comment by ciaranmcnulty** [comment]
It seems this type of error is not catchable https://gist.github.com/ciaranmcnulty/b1b8379fbafbdd76a50381a7382944ac

**Comment by jameshalsall** [comment]
We could possibly get around this by using reflection to ensure that the subject implements all of the methods in the interface and, if not, then offer to generate those stubs. If this sounds like something you want to pursue then I don't mind having a run at it, just let me know.

**Comment by ciaranmcnulty** [comment]
If you want to prototype it go for it. It's possible reflection will crap out first though.

**Comment by jameshalsall** [comment]
Yeah you're right, it's obviously just going to try and autoload the class by name using reflection. So the only other way would be manually parsing the content of the class file, which would be horrendous.

**Comment by ciaranmcnulty** [comment]
Possibly we could use https://github.com/Roave/BetterReflection but TBH there's probably a whole class of errors that aren't catchable.

Will have to improve the experience when an uncatchable error happens.

**Comment by jameshalsall** [comment]
That looks promising, I will take a look at that.

Currently if I have an interface `Foo`, and a class `Bar` that does not yet implement that interface, and my spec looks like this...

```
<?php

namespace spec;

use Bar;
use Foo;
use PhpSpec\ObjectBehavior;

class BarSpec extends ObjectBehavior
{
    function it_is_initializable()
    {
        $this->shouldHaveType(Bar::class);
    }

    function it_is_a_foo()
    {
        $this->shouldHaveType(Foo::class);
    }
}
```

Then I would expect PhpSpec to offer to update `Bar` for me so that it implements `Foo`. This workflow would remove the likelihood of encountering the issue in the first place. Currently the output would be

```
Bar
  17  - it has type
      expected an instance of Foo, but got [obj:Bar].
```

**Comment by ciaranmcnulty** [comment]
I'm concerned we'd be encouraging people to add methods without specs - perhaps we should offer to write a spec? ;-0

**Comment by jameshalsall** [comment]
I see what you're saying, but this would be the case where an interface already existed, something like:

1. Abstract type `Foo` already exists
2. We create a new concrete type called `Bar`
3. PhpSpec generates `Bar` for us
4. We update our `BarSpec` so that it says it should be a `Foo` type
5. PhpSpec then adds the `extends` or `implements` based on whether `Foo` is an abstract class or an interface, and any method stub.s

Between step 4 and 5 there is no state where the code would execute, so you would have to have empty method stubs in this example.
